### ISSUE TITLE
[Serve] Serve stalls out after repeated replica failures

### ISSUE THREAD

**Issue by spolcyn** [ray-project/ray#24419]
### What happened + What you expected to happen

1. After using a ServeHandle to repeatedly query a deployment whose replicas die a few times, the code stalls out and makes no further progress. Additionally, a `Task exception is never retrieved` exception is logged from within `asyncio`.

2. Expected behavior is the queries keep getting run, and returning `RayActorError` (or, in a real example, mostly the results)

3. See also https://github.com/ray-project/ray/issues/24228 which has a variety of other logs/`py-spy` output dumps

Example of the end of the output messages, after which no progress is made for at least 5min (in practice it hung until I killed it, which could be scores of minutes or hours -- the example below was left for 12+ hours, and remained hung)
```
(ServeController pid=97521) 2022-05-03 02:03:06,166     WARNING deployment_state.py:1402 -- Replica LongCall#UfgOLw of deployment LongCall failed health check, stopping it. component=serve deployment=LongCall replica=LongCall#UfgOLw
(Handler pid=97598) Task exception was never retrieved
(Handler pid=97598) future: <Task finished coro=<_wrap_awaitable() done, defined at /venv/lib/python3.7/asyncio/tasks.py:623> exception=RayActorError()>
(Handler pid=97598) Traceback (most recent call last):
(Handler pid=97598)   File "/venv/lib/python3.7/asyncio/tasks.py", line 630, in _wrap_awaitable
(Handler pid=97598)     return (yield from awaitable.__await__())
(Handler pid=97598) ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(Handler pid=97598)     class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
(Handler pid=97598)     actor_id: 7867eec6afde8ad702149a4b01000000
(Handler pid=97598)     pid: 97883
(Handler pid=97598)     name: SERVE_REPLICA::LongCall#UfgOLw
(Handler pid=97598)     namespace: serve
(Handler pid=97598)     ip: 172.17.0.3
(Handler pid=97598) The actor is dead because it was killed by `ray.kill`.
(ServeController pid=97521) 2022-05-03 02:03:06,372     INFO deployment_state.py:1211 -- Adding 1 replicas to deployment 'LongCall'. component=serve deployment=LongCall
```

The full output logs:
```
WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 67108864 bytes available. This will harm performance! You may be able to free up space by deleting files in /dev/shm. If you are inside a Docker container, you can increase /dev/shm size by passing '--shm-size=0.16gb' to 'docker run' (or add it to the run_options list in a Ray cluster config). Make sure to set this to more than 30% of available RAM.
(ServeController pid=97521) 2022-05-03 02:01:43,507     INFO checkpoint_path.py:15 -- Using RayInternalKVStore for controller checkpoint and recovery.
(ServeController pid=97521) 2022-05-03 02:01:43,609     INFO http_state.py:109 -- Starting HTTP proxy with name 'SERVE_CONTROLLER_ACTOR:SERVE_PROXY_ACTOR-node:172.17.0.3-0' on node 'node:172.17.0.3-0' listening on '127.0.0.1:8000'
2022-05-03 02:01:44,460 INFO api.py:795 -- Started detached Serve instance in namespace 'serve'.
(HTTPProxyActor pid=97561) INFO:     Started server process [97561]
(ServeController pid=97521) 2022-05-03 02:01:45,281     INFO deployment_state.py:1211 -- Adding 1 replicas to deployment 'LongCall'. component=serve deployment=LongCall
(Handler pid=97598) calling handle
(Handler pid=97598) 2022-05-03 02:01:45,270     INFO api.py:615 -- Updating deployment 'LongCall' to version 'v1'. component=serve deployment=LongCall
(Handler pid=97598) 2022-05-03 02:01:46,273     INFO api.py:631 -- Deployment 'LongCall:v1' is ready at `http://127.0.0.1:8000/LongCall`. component=serve deployment=LongCall
  0%|          | 0/5 [00:00<?, ?it/s]
(LongCall pid=97633) starting 0!
(ServeController pid=97521) 2022-05-03 02:01:51,317     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#WCHiiv after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:01:56,320     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#WCHiiv after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:01,324     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#WCHiiv after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:01,324     INFO deployment_state.py:559 -- Replica LongCall#WCHiiv failed the health check 3times in a row, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:01,324     WARNING deployment_state.py:1402 -- Replica LongCall#WCHiiv of deployment LongCall failed health check, stopping it. component=serve deployment=LongCall replica=LongCall#WCHiiv
(Handler pid=97598) Task exception was never retrieved
(Handler pid=97598) future: <Task finished coro=<_wrap_awaitable() done, defined at /venv/lib/python3.7/asyncio/tasks.py:623> exception=RayActorError()>
(Handler pid=97598) Traceback (most recent call last):
(Handler pid=97598)   File "/venv/lib/python3.7/asyncio/tasks.py", line 630, in _wrap_awaitable
(Handler pid=97598)     return (yield from awaitable.__await__())
(Handler pid=97598) ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(Handler pid=97598)     class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
(Handler pid=97598)     actor_id: ee5d03c2a30020098ef71ac001000000
(Handler pid=97598)     pid: 97633
(Handler pid=97598)     name: SERVE_REPLICA::LongCall#WCHiiv
(Handler pid=97598)     namespace: serve
(Handler pid=97598)     ip: 172.17.0.3
(Handler pid=97598) The actor is dead because it was killed by `ray.kill`.
(Handler pid=97598) Task exception was never retrieved
(Handler pid=97598) future: <Task finished coro=<_wrap_awaitable() done, defined at /venv/lib/python3.7/asyncio/tasks.py:623> exception=RayActorError()>
(Handler pid=97598) Traceback (most recent call last):
(Handler pid=97598)   File "/venv/lib/python3.7/asyncio/tasks.py", line 630, in _wrap_awaitable
(Handler pid=97598)     return (yield from awaitable.__await__())
(Handler pid=97598) ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(Handler pid=97598)     class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
(Handler pid=97598)     actor_id: ee5d03c2a30020098ef71ac001000000
(Handler pid=97598)     pid: 97633
(Handler pid=97598)     name: SERVE_REPLICA::LongCall#WCHiiv
(Handler pid=97598)     namespace: serve
(Handler pid=97598)     ip: 172.17.0.3
(Handler pid=97598) The actor is dead because it was killed by `ray.kill`.
 60%|██████    | 3/5 [00:15<00:10,  5.02s/it]
(ServeController pid=97521) 2022-05-03 02:02:01,531     INFO deployment_state.py:1211 -- Adding 1 replicas to deployment 'LongCall'. component=serve deployment=LongCall
(Handler pid=97598) getting futures
(Handler pid=97598) got an actor error! retrying 2
100%|██████████| 5/5 [00:16<00:00,  3.22s/it]
(LongCall pid=97699) starting 3!
(ServeController pid=97521) 2022-05-03 02:02:07,470     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#NVIBBE after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:12,481     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#NVIBBE after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:17,492     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#NVIBBE after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:17,492     INFO deployment_state.py:559 -- Replica LongCall#NVIBBE failed the health check 3times in a row, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:17,493     WARNING deployment_state.py:1402 -- Replica LongCall#NVIBBE of deployment LongCall failed health check, stopping it. component=serve deployment=LongCall replica=LongCall#NVIBBE
(Handler pid=97598) getting futures
(Handler pid=97598) got an actor error! retrying 4
(Handler pid=97598) Task exception was never retrieved
(Handler pid=97598) future: <Task finished coro=<_wrap_awaitable() done, defined at /venv/lib/python3.7/asyncio/tasks.py:623> exception=RayActorError()>
(Handler pid=97598) Traceback (most recent call last):
(Handler pid=97598)   File "/venv/lib/python3.7/asyncio/tasks.py", line 630, in _wrap_awaitable
(Handler pid=97598)     return (yield from awaitable.__await__())
(Handler pid=97598) ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(Handler pid=97598)     class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
(Handler pid=97598)     actor_id: ef7d78a1a199b8cc8e143bad01000000
(Handler pid=97598)     pid: 97699
(Handler pid=97598)     name: SERVE_REPLICA::LongCall#NVIBBE
(Handler pid=97598)     namespace: serve
(Handler pid=97598)     ip: 172.17.0.3
(Handler pid=97598) The actor is dead because it was killed by `ray.kill`.
(Handler pid=97598) Task exception was never retrieved
(Handler pid=97598) future: <Task finished coro=<_wrap_awaitable() done, defined at /venv/lib/python3.7/asyncio/tasks.py:623> exception=RayActorError()>
(Handler pid=97598) Traceback (most recent call last):
(Handler pid=97598)   File "/venv/lib/python3.7/asyncio/tasks.py", line 630, in _wrap_awaitable
(Handler pid=97598)     return (yield from awaitable.__await__())
(Handler pid=97598) ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(Handler pid=97598)     class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
(Handler pid=97598)     actor_id: ef7d78a1a199b8cc8e143bad01000000
(Handler pid=97598)     pid: 97699
(Handler pid=97598)     name: SERVE_REPLICA::LongCall#NVIBBE
(Handler pid=97598)     namespace: serve
(Handler pid=97598)     ip: 172.17.0.3
(Handler pid=97598) The actor is dead because it was killed by `ray.kill`.
(ServeController pid=97521) 2022-05-03 02:02:17,699     INFO deployment_state.py:1211 -- Adding 1 replicas to deployment 'LongCall'. component=serve deployment=LongCall
(Handler pid=97598) getting futures
(Handler pid=97598) got an actor error! retrying 2
(Handler pid=97598) getting futures
(Handler pid=97598) got an actor error! retrying 1
(LongCall pid=97760) starting 4!
(ServeController pid=97521) 2022-05-03 02:02:23,626     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#GuRmDb after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:28,633     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#GuRmDb after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:33,641     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#GuRmDb after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:33,642     INFO deployment_state.py:559 -- Replica LongCall#GuRmDb failed the health check 3times in a row, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:33,642     WARNING deployment_state.py:1402 -- Replica LongCall#GuRmDb of deployment LongCall failed health check, stopping it. component=serve deployment=LongCall replica=LongCall#GuRmDb
(Handler pid=97598) getting futures
(Handler pid=97598) got an actor error! retrying 1
(Handler pid=97598) Task exception was never retrieved
(Handler pid=97598) future: <Task finished coro=<_wrap_awaitable() done, defined at /venv/lib/python3.7/asyncio/tasks.py:623> exception=RayActorError()>
(Handler pid=97598) Traceback (most recent call last):
(Handler pid=97598)   File "/venv/lib/python3.7/asyncio/tasks.py", line 630, in _wrap_awaitable
(Handler pid=97598)     return (yield from awaitable.__await__())
(Handler pid=97598) ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(Handler pid=97598)     class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
(Handler pid=97598)     actor_id: 0f203e24912774f27d81279001000000
(Handler pid=97598)     pid: 97760
(Handler pid=97598)     name: SERVE_REPLICA::LongCall#GuRmDb
(Handler pid=97598)     namespace: serve
(Handler pid=97598)     ip: 172.17.0.3
(Handler pid=97598) The actor is dead because it was killed by `ray.kill`.
(Handler pid=97598) Task exception was never retrieved
(Handler pid=97598) future: <Task finished coro=<_wrap_awaitable() done, defined at /venv/lib/python3.7/asyncio/tasks.py:623> exception=RayActorError()>
(Handler pid=97598) Traceback (most recent call last):
(Handler pid=97598)   File "/venv/lib/python3.7/asyncio/tasks.py", line 630, in _wrap_awaitable
(Handler pid=97598)     return (yield from awaitable.__await__())
(Handler pid=97598) ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(Handler pid=97598)     class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
(Handler pid=97598)     actor_id: 0f203e24912774f27d81279001000000
(Handler pid=97598)     pid: 97760
(Handler pid=97598)     name: SERVE_REPLICA::LongCall#GuRmDb
(Handler pid=97598)     namespace: serve
(Handler pid=97598)     ip: 172.17.0.3
(Handler pid=97598) The actor is dead because it was killed by `ray.kill`.
(ServeController pid=97521) 2022-05-03 02:02:33,848     INFO deployment_state.py:1211 -- Adding 1 replicas to deployment 'LongCall'. component=serve deployment=LongCall
(Handler pid=97598) getting futures
(Handler pid=97598) got an actor error! retrying 4
(Handler pid=97598) getting futures
(Handler pid=97598) got an actor error! retrying 2
(LongCall pid=97822) starting 1!
(ServeController pid=97521) 2022-05-03 02:02:39,889     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#zjPEqo after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:44,893     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#zjPEqo after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:49,903     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#zjPEqo after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:49,903     INFO deployment_state.py:559 -- Replica LongCall#zjPEqo failed the health check 3times in a row, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:02:49,903     WARNING deployment_state.py:1402 -- Replica LongCall#zjPEqo of deployment LongCall failed health check, stopping it. component=serve deployment=LongCall replica=LongCall#zjPEqo
(Handler pid=97598) getting futures
(Handler pid=97598) got an actor error! retrying 1
(Handler pid=97598) Task exception was never retrieved
(Handler pid=97598) future: <Task finished coro=<_wrap_awaitable() done, defined at /venv/lib/python3.7/asyncio/tasks.py:623> exception=RayActorError()>
(Handler pid=97598) Traceback (most recent call last):
(Handler pid=97598)   File "/venv/lib/python3.7/asyncio/tasks.py", line 630, in _wrap_awaitable
(Handler pid=97598)     return (yield from awaitable.__await__())
(Handler pid=97598) ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(Handler pid=97598)     class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
(Handler pid=97598)     actor_id: 33d3a866333b1029d9dfa5e901000000
(Handler pid=97598)     pid: 97822
(Handler pid=97598)     name: SERVE_REPLICA::LongCall#zjPEqo
(Handler pid=97598)     namespace: serve
(Handler pid=97598)     ip: 172.17.0.3
(Handler pid=97598) The actor is dead because it was killed by `ray.kill`.
(Handler pid=97598) Task exception was never retrieved
(Handler pid=97598) future: <Task finished coro=<_wrap_awaitable() done, defined at /venv/lib/python3.7/asyncio/tasks.py:623> exception=RayActorError()>
(Handler pid=97598) Traceback (most recent call last):
(Handler pid=97598)   File "/venv/lib/python3.7/asyncio/tasks.py", line 630, in _wrap_awaitable
(Handler pid=97598)     return (yield from awaitable.__await__())
(Handler pid=97598) ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(Handler pid=97598)     class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
(Handler pid=97598)     actor_id: 33d3a866333b1029d9dfa5e901000000
(Handler pid=97598)     pid: 97822
(Handler pid=97598)     name: SERVE_REPLICA::LongCall#zjPEqo
(Handler pid=97598)     namespace: serve
(Handler pid=97598)     ip: 172.17.0.3
(Handler pid=97598) The actor is dead because it was killed by `ray.kill`.
(ServeController pid=97521) 2022-05-03 02:02:50,111     INFO deployment_state.py:1211 -- Adding 1 replicas to deployment 'LongCall'. component=serve deployment=LongCall
(Handler pid=97598) getting futures
(Handler pid=97598) got an actor error! retrying 4
(Handler pid=97598) getting futures
(Handler pid=97598) got an actor error! retrying 2
(LongCall pid=97883) starting 1!
(ServeController pid=97521) 2022-05-03 02:02:56,139     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#UfgOLw after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:03:01,146     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#UfgOLw after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:03:06,165     INFO deployment_state.py:494 -- Didn't receive health check response for replica LongCall#UfgOLw after 5.0s, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:03:06,165     INFO deployment_state.py:559 -- Replica LongCall#UfgOLw failed the health check 3times in a row, marking it unhealthy.
(ServeController pid=97521) 2022-05-03 02:03:06,166     WARNING deployment_state.py:1402 -- Replica LongCall#UfgOLw of deployment LongCall failed health check, stopping it. component=serve deployment=LongCall replica=LongCall#UfgOLw
(Handler pid=97598) Task exception was never retrieved
(Handler pid=97598) future: <Task finished coro=<_wrap_awaitable() done, defined at /venv/lib/python3.7/asyncio/tasks.py:623> exception=RayActorError()>
(Handler pid=97598) Traceback (most recent call last):
(Handler pid=97598)   File "/venv/lib/python3.7/asyncio/tasks.py", line 630, in _wrap_awaitable
(Handler pid=97598)     return (yield from awaitable.__await__())
(Handler pid=97598) ray.exceptions.RayActorError: The actor died unexpectedly before finishing this task.
(Handler pid=97598)     class_name: create_replica_wrapper.<locals>.RayServeWrappedReplica
(Handler pid=97598)     actor_id: 7867eec6afde8ad702149a4b01000000
(Handler pid=97598)     pid: 97883
(Handler pid=97598)     name: SERVE_REPLICA::LongCall#UfgOLw
(Handler pid=97598)     namespace: serve
(Handler pid=97598)     ip: 172.17.0.3
(Handler pid=97598) The actor is dead because it was killed by `ray.kill`.
(ServeController pid=97521) 2022-05-03 02:03:06,372     INFO deployment_state.py:1211 -- Adding 1 replicas to deployment 'LongCall'. component=serve deployment=LongCall
```

### Versions / Dependencies

Ray 1.12
Python 3.7.12
Ubuntu 20.04

### Reproduction script

Run the script below to reproduce -- typically takes a minute or two to get to the stalling/failure, but over multiple runs, it does get there for me.

```
import ray
import time
from ray import serve
from ray.cluster_utils import Cluster
from tqdm import tqdm

cluster = Cluster(
    initialize_head=True,
    head_node_args={
        "num_cpus": 16,
        "memory": 8 * (1024 ** 3),
    },
)

ray.init(namespace="serve")
serve.start(detached=True)


@serve.deployment(
    _health_check_period_s=2,
    _health_check_timeout_s=5,
    _autoscaling_config={
        "min_replicas": 1,
        "max_replicas": 5,
        "target_num_ongoing_requests_per_replica": 2,
    },
    version="v1",
    max_concurrent_queries=2,
)
class LongCall:
    def run(self, i: int):
        print(f"starting {i}!")
        time.sleep(60)
        print(f"done {i}!")
        return i


@ray.remote
class Handler:
    def handle(self):
        print("calling handle")
        LongCall.deploy()
        handle = LongCall.get_handle()

        future_to_index = {}
        for i in tqdm(range(5)):
            future = handle.run.remote(i)
            future_to_index[future] = i

        while True:
            print("getting futures")
            finished, _ = ray.wait(list(future_to_index.keys()), timeout=1.0)

            for fin in finished:
                try:
                    ray.get(fin)
                except ray.exceptions.RayActorError:
                    index = future_to_index.pop(fin)
                    print(f"got an actor error! retrying {index}")
                    future = handle.run.remote(index)
                    future_to_index[future] = index


h = Handler.remote()
ray.get(h.handle.remote())
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by spolcyn** [comment]
@simon-mo See this cleaned-up issue with a repro script for the earlier problem we were talking about

**Comment by simon-mo** [comment]
Thanks @spolcyn. This is a very detailed repro. I'll take a look.

**Comment by Gyubin** [comment]
Hi @spolcyn @simon-mo 
I had the same situation. The way I solved it is to increase the `_health_check_timeout_s` value to big integer(like 100).
The health_check_function seems to take a long time, so the ServeController keeps killing existing actors and adding new one.

**Comment by spolcyn** [comment]
@Gyubin Thanks for posting that mitigation here -- we've done a similar thing that seems to help (likely because there are fewer failures), and we've also put some of our longer-running calls (e.g., large inference batches) in threads to avoid blocking the health check.

@simon-mo -- on that topic (and happy to make another issue for this), it seems synchronous code running in Serve Deployments for time `t > health check timeout` will cause health check failures and then spurious kills (as it blocks the event loop). Is there a way to surface this limitation better in docs/warning messages, or even better run that synchronous code in an executor so it doesn't block the `asyncio` event loop in the actor?

For this particular issue, since our replicas take very little time to spinup though, I think the root cause of why the code stalls and doesn't make progress/raise exceptions remains though.

**Comment by simon-mo** [comment]
Ah @spolcyn  @Gyubin, thank you for these nice clue! I think it's straightforward to fix here to put healthcheck in a separate group of executor in the replica.

**Comment by spolcyn** [comment]
@simon-mo We've verified the fix with the repro script and have been waiting on 1.13.0 to see if the issue reoccurs on our production workloads.

Since we don't know exactly when 1.13.0 will be out though, feel free to close this issue in the meantime. We can open a new issue with a new repro later if we run into the same error for another reason.

**Comment by simon-mo** [comment]
Awesome. 1.13.0 is actively in the work and going through extensive testing process now. Closing this for now.

**Comment by spolcyn** [comment]
@simon-mo As it turns out, this didn't resolve the core problem, which appears to be in Ray Core (we've continued to run into the issue with Ray 1.13 and using just actors) -- posted a discussion thread here: https://discuss.ray.io/t/ray-actor-remote-call-blocks-forever/6820
